src/Perspectives/immutable/ungoverned/cluster/ClusterConservativeWithRecognizedCollateralsPerspective.sol

```diff
@@ -0,0 +1,145 @@
+
+            // cluster vaults must not have a hook target nor any operations disabled
+            (address hookTarget, uint32 hookedOps) = IEVault(vault).hookConfig();
+            testProperty(hookTarget == address(0), ERROR__HOOK_TARGET);
+            testProperty(hookedOps == 0, ERROR__HOOKED_OPS);
+        }
+
+        // cluster vaults must not have any config flags set
+        testProperty(IEVault(vault).configFlags() == 0, ERROR__CONFIG_FLAGS);
+
+        // cluster vaults must have certain liquidation cool off time
+        testProperty(IEVault(vault).liquidationCoolOffTime() == 1, ERROR__LIQUIDATION_COOL_OFF_TIME);
+
+        // cluster vaults must point to an ungoverned EulerRouter instance deployed by the factory
+        address oracle = IEVault(vault).oracle();
+        testProperty(routerFactory.isValidDeployment(oracle), ERROR__ORACLE);
+        testProperty(IEulerRouter(oracle).governor() == address(0), ERROR__ORACLE);
+
+        // Verify the unit of account is either USD or WETH
+        address unitOfAccount = IEVault(vault).unitOfAccount();
+        testProperty(unitOfAccount == USD || unitOfAccount == ETH || unitOfAccount == WETH, ERROR__UNIT_OF_ACCOUNT);
+
+        // the router must contain a valid pricing configuration
+        address fallbackOracle = IEulerRouter(oracle).fallbackOracle();
+        {
+            testProperty(fallbackOracle == address(0) || routerFactory.isValidDeployment(fallbackOracle), ERROR__ORACLE);
+            (,,, address resolvedOracle) = IEulerRouter(oracle).resolveOracle(1e18, vault, unitOfAccount);
+            testProperty(
+                (fallbackOracle != address(0) && resolvedOracle == fallbackOracle)
+                    || adapterRegistry.isValidAdapter(resolvedOracle, block.timestamp),
+                ERROR__ORACLE_ASSET
+            );
+        }
+
+        // cluster vaults must have collaterals set up
+        address[] memory ltvList = IEVault(vault).LTVList();
+        testProperty(ltvList.length > 0 && ltvList.length <= 10, ERROR__LTV_LENGTH);
+
+        // cluster vaults must have recognized collaterals with LTV set in range
+        for (uint256 i = 0; i < ltvList.length; ++i) {
+            address collateral = ltvList[i];
+            {
+                (,,, address resolvedOracle) = IEulerRouter(oracle).resolveOracle(1e18, vault, unitOfAccount);
+                testProperty(
+                    (fallbackOracle != address(0) && resolvedOracle == fallbackOracle)
+                        || adapterRegistry.isValidAdapter(resolvedOracle, block.timestamp),
+                    ERROR__ORACLE_COLLATERAL
+                );
+            }
+
+            // cluster vaults must have liquidation discount in a certain range
+            uint16 maxLiquidationDiscount = IEVault(vault).maxLiquidationDiscount();
+            testProperty(
+                maxLiquidationDiscount >= 0.05e4 && maxLiquidationDiscount <= 0.2e4, ERROR__LIQUIDATION_DISCOUNT
```

Comment:
Move these magic numbers to contract-level constants i.e. `MAX_LIQUIDATION_DISCOUNT_LOWER_BOUND`.

Highlighting a single instance, but there are several other instances throughout.

Reply:
I began to do this, and it doesn't seem like a good idea after all. For some parameters, you want the bound to be inclusive; for others, you want it to be exclusive. Some parameters require both bands; some parameters require just one of them. It became a mess. Maybe having magic numbers is more confusing at first glance, but overall, I think it's safer if those values are specified inline.